File: lwk_bindings/tests/bindings/amp0-setup.py
from lwk import *
import os


gitlab_ci = os.environ.get("GITLAB_CI")
ci_job_name = os.environ.get("CI_JOB_NAME")
ci_branch_name = os.environ.get("CI_COMMIT_BRANCH")
if gitlab_ci == "true":
    # We are in a CI job
    if ci_branch_name != "master" or ci_job_name != "amp0":
        print("Skipping test")
        quit()


# ANCHOR: amp0-setup
# Create signer and watch only credentials
username = "<username>";
password = "<password>";
mnemonic = "<mnemonic>";

mnemonic = Mnemonic.from_random(12); # ANCHOR: ignore
network = Network.testnet()
signer = Signer(mnemonic, network)
username = f"user{signer.fingerprint()}" # ANCHOR: ignore
password = f"pass{signer.fingerprint()}" # ANCHOR: ignore

# Collect signer data
signer_data = signer.amp0_signer_data();

# Connect to AMP0
amp0 = Amp0Connected(network, signer_data);

# Obtain and sign the authentication challenge
challenge = amp0.get_challenge();
sig = signer.amp0_sign_challenge(challenge);

# Login
amp0 = amp0.login(sig);

# Create a new AMP0 account
pointer = amp0.next_account();
account_xpub = signer.amp0_account_xpub(pointer);
amp_id = amp0.create_amp0_account(pointer, account_xpub);

# Create watch only entries
amp0.create_watch_only(username, password);

# Use watch only credentials to interact with AMP0
amp0 = Amp0(network, username, password, amp_id);
# ANCHOR_END: amp0-setup

File: lwk_bindings/go/amp0-setup.go
package main

import (
	"fmt"
	"log"
	"os"

	"lwk-go/lwk"
)

func main() {
	gitlabCI := os.Getenv("GITLAB_CI")
	ciJobName := os.Getenv("CI_JOB_NAME")
	ciBranchName := os.Getenv("CI_COMMIT_BRANCH")

	if gitlabCI == "true" {
		// We are in a CI job
		if ciBranchName != "master" || ciJobName != "amp0" {
			fmt.Println("Skipping test")
			return // Exits gracefully like quit()
		}
	}

	// ANCHOR: amp0-setup
	// Create signer and watch only credentials
	username := "<username>"
	password := "<password>"
	mnemonic, err := lwk.NewMnemonic("<mnemonic>")
	mnemonic, err = lwk.MnemonicFromRandom(12) // ANCHOR: ignore
	if err != nil {
		log.Fatal(err)
	}

	network := lwk.NetworkTestnet()
	signer, err := lwk.NewSigner(mnemonic, network)
	if err != nil {
		log.Fatal(err)
	}

	fingerprint, err := signer.Fingerprint()
	if err != nil {
		log.Fatal(err)
	}

	username = fmt.Sprintf("user%s", fingerprint) // ANCHOR: ignore
	password = fmt.Sprintf("pass%s", fingerprint) // ANCHOR: ignore

	// Collect signer data
	signerData, err := signer.Amp0SignerData()
	if err != nil {
		log.Fatal(err)
	}

	// Connect to AMP0
	amp0, err := lwk.NewAmp0Connected(network, signerData)
	if err != nil {
		log.Fatal(err)
	}

	// Obtain and sign the authentication challenge
	challenge, err := amp0.GetChallenge()
	if err != nil {
		log.Fatal(err)
	}
	sig, err := signer.Amp0SignChallenge(challenge)
	if err != nil {
		log.Fatal(err)
	}

	// Login
	amp0Connected, err := amp0.Login(sig)
	if err != nil {
		log.Fatal(err)
	}

	// Create a new AMP0 account
	pointer, err := amp0Connected.NextAccount()
	if err != nil {
		log.Fatal(err)
	}
	accountXpub, err := signer.Amp0AccountXpub(pointer)
	if err != nil {
		log.Fatal(err)
	}
	ampId, err := amp0Connected.CreateAmp0Account(pointer, accountXpub)
	if err != nil {
		log.Fatal(err)
	}

	// Create watch only entries
	if err := amp0Connected.CreateWatchOnly(username, password); err != nil {
		log.Fatal(err)
	}

	// Use watch only credentials to interact with AMP0
	amp0New, err := lwk.NewAmp0(network, username, password, ampId)
	if err != nil {
		log.Fatal(err)
	}
	// ANCHOR_END: amp0-setup
	_ = amp0New

}
